spyderlib/otherplugins.py

```diff
@@ -8,52 +8,71 @@
 import sys
 import traceback
 
 # Local imports
 from spyderlib.utils import programs
-
+from spyderlib.baseconfig import get_conf_path
+from spyderlib.utils.external.path import path as Path
 
 # Calculate path to `spyderplugins` package, where Spyder looks for all 3rd
 # party plugin modules
 PLUGIN_PATH = None
 if programs.is_module_installed("spyderplugins"):
     import spyderplugins
-    PLUGIN_PATH = osp.abspath(spyderplugins.__path__[0])
-    if not osp.isdir(PLUGIN_PATH):
+    PLUGIN_PATH = Path(spyderplugins.__path__[0]).abspath()
+    if not PLUGIN_PATH.isdir():
         # py2exe/cx_Freeze distribution: ignoring extra plugins
         PLUGIN_PATH = None
 
+USERPLUGIN_PATH = Path(get_conf_path("userplugins"))
+USERPLUGIN_PATH.makedirs_p()
+sys.path.append(USERPLUGIN_PATH)
+
 
-def get_spyderplugins(prefix, extension):
+def _get_spyderplugins(plugin_path, prefix):
     """Scan directory of `spyderplugins` package and
-    return the list of module names matching *prefix* and *extension*"""
+    return the list of module names matching *prefix*"""
     plist = []
     if PLUGIN_PATH is not None:
-        for name in os.listdir(PLUGIN_PATH):
-            modname, ext = osp.splitext(name)
-            if prefix is not None and not name.startswith(prefix):
+        for dirname in plugin_path.dirs(pattern=prefix + "*"):
+            if not (dirname / "__init__.py").isfile():
                 continue
-            if extension is not None and ext != extension:
-                continue
-            plist.append(modname)
+            plist.append(dirname.name)
+        for name in plugin_path.files(pattern=prefix + "*.py"):
+            plist.append(name.namebase)
     return plist
 
 
-def get_spyderplugins_mods(prefix, extension):
-    """Import modules that match *prefix* and *extension* from
+def _import_plugin(name, modlist):
+    """Import the plugin `modname` and add it to `modlist`"""
+    try:
+        __import__(name)
+        modlist.append(sys.modules[name])
+    except Exception:
+        sys.stderr.write(
+            "ERROR: 3rd party plugin import failed for `%s`\n" % name)
+        traceback.print_exc(file=sys.stderr)
+
+
+def get_spyderplugins_mods(prefix):
+    """Import modules that match *prefix* from
     `spyderplugins` package and return the list"""
     modlist = []
-    for modname in get_spyderplugins(prefix, extension):
-        name = 'spyderplugins.%s' % modname
-        try:
-            __import__(name)
-            modlist.append(sys.modules[name])
-        except Exception:
-            sys.stderr.write(
-                "ERROR: 3rd party plugin import failed for `%s`\n" % modname)
-            traceback.print_exc(file=sys.stderr)
+
+    # Load user plugins
+    user_plugins = _get_spyderplugins(USERPLUGIN_PATH, prefix)
+    sys.path.insert(0, USERPLUGIN_PATH)
+    try:
+        for modname in user_plugins:
+            _import_plugin(modname, modlist)
+    finally:
+        del sys.path[0]

```

Comment:
I think if you add something to `sys.path` temporarily, you should definitely remove it. Maybe it's also a good idea to test if you remove the correct element (in case something else got prepended on `sys.path` in the meantime, maybe like so: 

```
if USERPLUGIN_PATH in sys.path:
    sys.path.remove(USERPLUGIN_PATH)
```
(see [here](http://stackoverflow.com/a/2793356/599884))